parsl/executors/high_throughput/mpi_executor.py

```diff
@@ -0,0 +1,105 @@
+from parsl.executors.status_handling import BlockProviderExecutor
+from parsl.jobs.states import JobStatus
+from parsl.providers import LocalProvider
+from parsl.providers.base import ExecutionProvider
+
+
+class MPIExecutor(HighThroughputExecutor):
+    """A version of :class:`~parsl.HighThroughputExecutor` tuned for executing MPI tasks
+
+    Parameters
+    ----------
+    max_tasks: int
+        Maximum number of MPI applications to run at once per block
+    """
+
+    @typeguard.typechecked
+    def __init__(self,
+                 label: str = 'MPIExecutor',
+                 provider: ExecutionProvider = LocalProvider(),
+                 launch_cmd: Optional[str] = None,
+                 address: Optional[str] = None,
+                 worker_ports: Optional[Tuple[int, int]] = None,
+                 worker_port_range: Optional[Tuple[int, int]] = (54000, 55000),
+                 interchange_port_range: Optional[Tuple[int, int]] = (55000, 56000),
+                 storage_access: Optional[List[Staging]] = None,
+                 working_dir: Optional[str] = None,
+                 worker_debug: bool = False,
+                 max_tasks: int = 1,
+                 prefetch_capacity: int = 0,
+                 heartbeat_threshold: int = 120,
+                 heartbeat_period: int = 30,
+                 drain_period: Optional[int] = None,
+                 poll_period: int = 10,
+                 address_probe_timeout: Optional[int] = None,
+                 worker_logdir_root: Optional[str] = None,
+                 mpi_launcher: str = "mpiexec",
+                 block_error_handler: Union[bool, Callable[[BlockProviderExecutor, Dict[str, JobStatus]], None]] = True,
+                 encrypted: bool = False):
+        super().__init__(
+            # Hard-coded settings
+            cores_per_worker=1e-9,  # Ensures there will always be enough workers
+            enable_mpi_mode=True,
+            max_workers_per_node=max_tasks,
+
+            # Everything else
+            label=label,
+            provider=provider,
+            launch_cmd=launch_cmd,
+            address=address,
+            worker_ports=worker_ports,
+            worker_port_range=worker_port_range,
+            interchange_port_range=interchange_port_range,
+            storage_access=storage_access,
+            working_dir=working_dir,
+            worker_debug=worker_debug,
+            prefetch_capacity=prefetch_capacity,
+            heartbeat_threshold=heartbeat_threshold,
+            heartbeat_period=heartbeat_period,
+            drain_period=drain_period,
+            poll_period=poll_period,
+            address_probe_timeout=address_probe_timeout,
+            worker_logdir_root=worker_logdir_root,
+            mpi_launcher=mpi_launcher,
+            block_error_handler=block_error_handler,
+            encrypted=encrypted
+        )
+
+
+# Update the docstring on import
+def _update_from_htex_docstring():
```

Comment:
:exploding_head:

Reply:
Do you hate this? I felt a little foolish writing it, but don't think any Python documentation tools have a "get parameter definitions from the superclass" function

Reply:
I really dislike this and have been trying to think of nicer ways to do this.

One way I came up with is to stop trying to regard MPIExecutor as "real HTEX but different", which leads to this _update_from_htex_docstring, and instead view htex as existing in two equal first order modes: mpi and not-mpi.

Then theres a shared docstring, which you could define (in HighThroughputExecutor) as  _shared_docstring= """
...
"""

and then the HTEX docstring is  `__doc__ = """whatever non-mpi htex intro""" + _shared_docstring
and the MPI docstring is "__doc__ = """whatever mpi htex intro""" + _shared_docstring

That removes any parsing from this function, by making the docstrings already separated in source into three chapters, to be combined using `+`.

Reply:
Good idea! I implemented something like that, after seeing how to set docstrings programmatically https://stackoverflow.com/questions/70863543/can-a-python-docstring-be-calculated-f-string-or-expression